src/uct/tcp/tcp_ep.c

```diff
@@ -947,37 +932,38 @@ ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *hea
 
     iface->outstanding += ep->tx.length;
 
-    status = ucs_socket_sendv_nb(ep->fd, ctx_iov, ctx->iov_cnt,
+    status = ucs_socket_sendv_nb(ep->fd, ctx->iov, ctx->iov_cnt,
                                  &ep->tx.offset, NULL, NULL);
 
     ucs_trace_data("tcp_ep %p: sendv %zu bytes", ep, ep->tx.offset);
 
     iface->outstanding -= ep->tx.offset;
-    if ((ep->tx.offset != ep->tx.length) &&
-        ((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS))) {
-        ctx->comp     = comp;
-        ep->ctx_caps |= UCS_BIT(UCT_TCP_EP_CTX_TYPE_ZCOPY_TX);
-
-        if ((header_length != 0) &&
-            /* check whether a user's header was sent or not */
-            (ep->tx.offset < (sizeof(*hdr) + header_length))) {
-            ucs_assert(header_length <= iface->config.zcopy.max_hdr);
-            /* if the user's header wasn't sent completely, copy it to the EP
-             * TX buffer (after Zcopy context and IOVs) for retransmission */
-            ctx_iov[1].iov_base = ep->tx.buf + iface->config.zcopy.hdr_offset;
-            memcpy(ctx_iov[1].iov_base, header, header_length);
-        }
-
-        ctx->iov_index  = 0;
-        ctx->iov_offset = 0;
+    if ((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS)) {
+        UCT_TL_EP_STAT_OP(&ep->super, AM, ZCOPY, hdr->length);
+
+        if (ep->tx.offset != ep->tx.length) {
+            ctx->comp     = comp;
```

Comment:
maybe extract this code to another function?
like uct_tcp_ep_set_outstanding_zcopy

Reply:
done